cmd/bugzilla-backporter/main.go

```diff
@@ -0,0 +1,187 @@
+package main
+
+import (
+	"encoding/json"
+	"flag"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"os"
+	"strconv"
+	"time"
+
+	"github.com/prometheus/client_golang/prometheus"
+	log "github.com/sirupsen/logrus"
+	"k8s.io/test-infra/prow/bugzilla"
+	prowConfig "k8s.io/test-infra/prow/config"
+	"k8s.io/test-infra/prow/interrupts"
+	"k8s.io/test-infra/prow/metrics"
+	"k8s.io/test-infra/prow/pjutil"
+)
+
+var (
+	bzbpMetrics = struct {
+		httpRequestDuration *prometheus.HistogramVec
+		httpResponseSize    *prometheus.HistogramVec
+		errorRate           *prometheus.CounterVec
+	}{
+		httpRequestDuration: prometheus.NewHistogramVec(
+			prometheus.HistogramOpts{
+				Name:    "bugzilla_backporter_http_request_duration_seconds",
+				Help:    "http request duration in seconds",
+				Buckets: []float64{0.0005, 0.001, 0.0025, 0.005, 0.01, 0.025, 0.05, 0.1, 0.25, 0.5, 1, 2},
+			},
+			[]string{"status", "path"},
+		),
+		httpResponseSize: prometheus.NewHistogramVec(
+			prometheus.HistogramOpts{
+				Name:    "bugzilla_backporter_http_response_size_bytes",
+				Help:    "http response size in bytes",
+				Buckets: []float64{256, 512, 1024, 2048, 4096, 6144, 8192, 10240, 12288, 16384, 24576, 32768, 40960, 49152, 57344, 65536},
+			},
+			[]string{"status", "path"},
+		),
+		errorRate: prometheus.NewCounterVec(
+			prometheus.CounterOpts{
+				Name: "bugzilla_backporter_error_rate",
+				Help: "number of errors, sorted by label/type",
+			},
+			[]string{"error"},
+		),
+	}
+)
+
+func recordError(label string) {
+	labels := prometheus.Labels{"error": label}
+	bzbpMetrics.errorRate.With(labels).Inc()
+}
+
+type options struct {
+	configPath   string
+	registryPath string
+	prowPath     string
+	jobPath      string
+	logLevel     string
+	address      string
+	uiAddress    string
+	gracePeriod  time.Duration
+	cycle        time.Duration
+	validateOnly bool
+	flatRegistry bool
+}
+
+type traceResponseWriter struct {
+	http.ResponseWriter
+	statusCode int
+	size       int
+}
+
+func handleWithMetrics(h http.HandlerFunc) http.HandlerFunc {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		t := time.Now()
+		// Initialize the status to 200 in case WriteHeader is not called
+		trw := &traceResponseWriter{ResponseWriter: w, statusCode: http.StatusOK}
+		h(trw, r)
+		latency := time.Since(t)
+		labels := prometheus.Labels{"status": strconv.Itoa(trw.statusCode), "path": r.URL.EscapedPath()}
+		bzbpMetrics.httpRequestDuration.With(labels).Observe(latency.Seconds())
+		bzbpMetrics.httpResponseSize.With(labels).Observe(float64(trw.size))
+	})
+}
+
+// Needs additional work
+func getAuthToken() []byte {
```

Comment:
I am not sure I understand, could you please explain what you mean by that?

Reply:
Commandline flags so this can be configured to read the token from a file. It seems we already have some helpers for this in upstream: https://github.com/kubernetes/test-infra/blob/4b5c7c99a851eb427f5c77bd0c8d11526f7b63c4/prow/cmd/hook/main.go#L57-L85